sign/ed25519/ed25519.go

```diff
@@ -2,42 +2,93 @@ package ed25519
 
 import (
 	"bytes"
+	"crypto"
+	"crypto/rand"
 	"crypto/sha512"
 	"encoding/binary"
+	"errors"
+	"io"
 	"math/bits"
+	"strconv"
 
 	fp "github.com/cloudflare/circl/math/fp25519"
 )
 
 // Size is the length in bytes of Ed25519 keys.
 const Size = fp.Size
 
-// PubKey represents a public key of Ed25519.
-type PubKey [Size]byte
+// PublicKey represents a public key of Ed25519.
+type PublicKey [Size]byte
 
-// PrivKey represents a private key of Ed25519.
-type PrivKey [Size]byte
+// PrivateKey represents a private key of Ed25519. Implements crypto.Signer (golang.org/pkg/crypto/#Signer) interface.
+type PrivateKey struct {
+	public PublicKey  // This is a copy of the corresponding public key.
+	seed   [Size]byte // This is the secret (seed) key.
+}
 
 // Signature represents an Ed25519 signature.
 type Signature [2 * Size]byte
 
-// Pure corresponds to the Ed25519 PureEdDSA instance.
-type Pure struct{}
+// GetPublic returns a PublicKey corresponding to the private key.
+func (priv *PrivateKey) GetPublic() *PublicKey {
+	var public PublicKey
+	copy(public[:], priv.public[:])
+	return &public
+}
+
+// Public returns a crypto.PublicKey corresponding to the private key.
+func (priv *PrivateKey) Public() crypto.PublicKey { return priv.GetPublic() }
+
+// Sign signs the given message with priv.
+// Ed25519 performs two passes over messages to be signed and therefore cannot
+// handle pre-hashed messages. Thus opts.HashFunc() must return zero to
+// indicate the message hasn't been hashed. This can be achieved by passing
+// crypto.Hash(0) as the value for opts.
+func (priv *PrivateKey) Sign(rand io.Reader, message []byte, opts crypto.SignerOpts) ([]byte, error) {
+	if opts.HashFunc() != crypto.Hash(0) {
+		return nil, errors.New("ed25519: cannot sign hashed message")
+	}
+	sig := Sign(priv, message)
+	return sig[:], nil
+}
+
+// GenerateKey generates a pair of keys.
+func GenerateKey(rnd io.Reader) (*PrivateKey, *PublicKey, error) {
+	if rnd == nil {
+		rnd = rand.Reader
+	}
+
+	var seed [Size]byte
+	if _, err := io.ReadFull(rnd, seed[:]); err != nil {
+		return nil, nil, err
+	}
+	private := NewKeyFromSeed(seed[:])
+	return private, private.Public().(*PublicKey), nil
+}
 
-// KeyGen generates a public key from a secret key.
-func (e Pure) KeyGen(public *PubKey, private *PrivKey) {
-	k := sha512.Sum512(private[:])
+// NewKeyFromSeed calculates a private key from a seed. It will panic if
+// len(seed) is not SeedSize. This function is provided for interoperability
+// with RFC 8032. RFC 8032's private keys correspond to seeds in this
+// package.
+func NewKeyFromSeed(seed []byte) *PrivateKey {
+	if l := len(seed); l != Size {
+		panic("ed25519: bad seed length: " + strconv.Itoa(l))
+	}
+	var P pointR1
+	var private PrivateKey
+	k := sha512.Sum512(seed)
```

Comment:
Perhaps rename seed -> secret?
Following https://ed25519.cr.yp.to/ed25519-20110926.pdf and https://tools.ietf.org/html/rfc8032#section-3.2,
the input for this function is the secret key `k` (the EdDSA private key). The result of hashing `k` is internally used for signing/verification and referred to as `s`.

Reply:
Since this is an internal variable I can rename it to `private` key. 
I was trying to follow the stdlib implementation, but `private` key is the correct name.